File: Data_Structures/Singly_Linked_list.js
/**
 * Linked List --- It is a orders list of values without any index 
 *              - It has three properties
 *                      -Head
 *                      -Length
 *                      -Tail
 *              - Every element contains a value and a pointer to the next element
 *              - Each element is called a node 
 *              - And each node points to another node or NULL
 *              - Random excess to an element is not allowed
 * 
 *          - Really good for insertion and deletion 
 * 
 *      -- Singly linked list - only has uni-directional connection
*/

class node {
    constructor(val) {
        this.value = val;
        this.next = null;
    }
}

class singly_linked_list {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    push(val) {                          //Pushing an element to the last of the linked list
        var element = new node(val);
        if (this.head == null) {
            this.head = element;
            this.tail = element;
            this.tail.next = null;
        } else {
            this.tail.next = element;
            this.tail = element;
        }
        this.length++;
        return this;
    }

    pop() {                               // Removing an element from the last of the linked list
        var element_1 = this.head;
        if (this.head == null) {
            return undefined;
        }
        if (this.length == 1) {
            this.head = null;
            this.tail = null;
            this.length = 0;
            return this;
        }
        this.length--;
        var element_2 = this.head.next;
        while (element_2.next != null) {
            element_1 = element_2;
            element_2 = element_2.next;
        }
        this.tail = element_1;
        this.tail.next = null;
        return this;
    }

    shift() {                                    //Removing an element from the first of the linked list
        if (this.length == 0) {
            return undefined;
        }
        if (this.length == 1) {
            this.head = null;
            this.tail = null;
            this.length = 0;
            return this;
        }
        this.head = this.head.next;
        this.length--;
        return this;
    }

    unshift(val) {                         //Adds an element to the first of the linked list
        if(this.length == 0){
            this.push(val);
            return this;
        }
        this.length++;
        var cur_element = this.head;
        var new_head = new node(val);
        new_head.next = cur_element;
        this.head = new_head;
        return this;
    }
    get_node(index){
        if (this.head == null || index < 0 || index >= this.length){
            return undefined;
        }
        var current = this.head;
        for (let i = 0; i < index; i++) {
            current = current.next;
        }
        return current;
    }

    set_value(index, value){
        if(this.head == null || index < 0 || index >= this.length){
            return false;
        }
        var current = this.head;
        for (let i = 0; i < index; i++){
            current = current.next;
        }
        current.value = value;
        return true;
    }

    insert(index, value){
        if(index < 0 || index > this.length){
            return false;
        }
        if (index == 0){
            this.unshift(value);
            this.tail.next = null;
            return true;
        }
        if(index == this.length){
            this.push(value);
            return true;
        }
        var prev_element = this.get_node(index - 1);
        var next_element = prev_element.next;
        var new_element = new node(value);
        prev_element.next = new_element;
        new_element.next = next_element;
        this.length++;
        return true;
    }

    delete(index){
        if (index <0 || index >= this.length){
            return false
        }
        if(index == 0){
            this.shift();
            return true;
        }
        var prev_element = this.get_node(index - 1);
        var next_element = prev_element.next.next;
        prev_element.next = next_element;
        this.length--;
        return true;
    }

    reverse(){
        console.log("-----");
    }

    print() {
        var cur_element = this.head;
        for(;cur_element != null;cur_element = cur_element.next) {
            console.log(cur_element.value);
        }
        return this;
    }
}

var list = new singly_linked_list();

list.push(1);
list.push(2);
list.push(3);
list.push(4);
list.push(5);

console.log(list);

File: Data_Structures/Singly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct node {
    int value;
    struct node *next;
};

struct linked_list{
    struct node *head;
    struct node *tail;
    int length; 
};

void push(struct linked_list *list, int value){
    struct node *new_node = (struct node *)malloc(sizeof(struct node));
    new_node->value = value;
    new_node->next = NULL;
    if(list->length == 0){
        list->head = new_node;
        list->tail = new_node;
    }else{
        list->tail->next = new_node;
        list->tail = new_node;
    }
    list->length++;
    printf("pushed %d\n", value);
}

void pop(struct linked_list *list){
    if(list->length == 0){
        printf("list is empty\n");
    }else{
        if(list->length == 1){
            list->head = NULL;
            list->tail = NULL;
        }else{
            struct node *temp = list->head;
            while(temp->next != list->tail){
                temp = temp->next;
            }
            temp->next = NULL;
            list->tail = temp;
        }
        list->length--;
    }
    printf("popped\n");
}

void print_list(struct linked_list* list){
    struct node *current = list->head;
    for(;current != NULL; current = current->next){
        printf("%d->", current->value);
    }
    printf("NULL Length==%d\n", list->length);
}


void main(){
    struct linked_list* list = (struct linked_list*)malloc(sizeof(struct linked_list));
    list->tail = NULL;
    list->head = NULL;
    list->length = 0;
    push(list, 1);
    push(list, 13);
    push(list, 3);
    push(list, 5);
    push(list, 7);
    print_list(list);
    pop(list);
    print_list(list);
    // push(list, 1);
}
